### PR TITLE
E2.02/adapt notification per project

### PR DESCRIPTION
**Pull request by wow-such-code**
* Moves creation of E-Mail Content out of the use case
* Adds NotificationContent DTO object used by...
* ...newly implemented use case, that collects the information for each project and subscriber, which samples have been updated

### CODE REVIEW

**Comment by wow-such-code** [comment]
tests are not working yet due to the pom not being set up correctly (we think), but feedback is appreciated!

File: README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 This CLI is intended to search the database for projects (samples of a project)
 that have received a new status. This is only done for the current day the tool is executed.
 
-People who subscribed to be notified about their projects changes
+People who are subscribed to be notified about their projects' changes
```
Comment (jenniferboedker) [inline README.md:5]:
```suggestion
People who are subscribed are notified about their projects' changes
```

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy [hunk 2]
```diff
@@ -21,10 +22,11 @@ class CreateNotificationPresenter implements CreateNotificationOutput, FetchSubs
...
 
     @Override
-    void createdNotifications(Map<Subscriber, String> notificationPerSubscriber) {
-        notificationPerSubscriber.each {
-            this.notificationPerSubscriber.put(it.key, it.value)
-        }
+    void createdNotifications(List<NotificationContent> notifications) {
+      //TODO
+//        notificationPerSubscriber.each {
+//            this.notificationPerSubscriber.put(it.key, it.value)
```
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:28]: What are you missing here? Is this were Steffens implementation comes into play?

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/FetchUpdatedDataSource.groovy [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import life.qbic.business.subscription.Subscriber
+
+import java.time.LocalDate
+
+/**
+ * <p>Fetches the subscribers from the database</p>
+ *
+ * @since 1.0.0
+ */
+interface FetchUpdatedDataSource {
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/FetchUpdatedDataSource.groovy:14]: What is the idea of this interface? Why did you add it? Was there a method missing in the old interface? Or in which ways does it differ to the other DS?

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy [hunk 4]
```diff
@@ -21,10 +22,11 @@ class CreateNotificationPresenter implements CreateNotificationOutput, FetchSubs
     }
 
     @Override
-    void createdNotifications(Map<Subscriber, String> notificationPerSubscriber) {
-        notificationPerSubscriber.each {
-            this.notificationPerSubscriber.put(it.key, it.value)
-        }
+    void createdNotifications(List<NotificationContent> notifications) {
+      //TODO
```
Comment (wow-such-code) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:26]: Not sure if this is to be tackled in another branch/by somebody else

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 5]
```diff
@@ -15,156 +19,100 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
+//    /**
+//     * Sends failure notifications that have been
+//     * recorded during the use case.
+//     * @param notification containing a failure message
+//     * @since 1.0.0
+//     */
+//    @Override
+//    void failNotification(String notification) {
+//        output.failNotification(notification)
+//    }
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:117]: You can remove this I think :)

```suggestion
```

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 6]
```diff
@@ -15,156 +19,100 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
+            
+            //4. get subscribers of this projects
+            List<Subscriber> subscribers = getSubscriberForProject(projectCode)
+            
+            for(Subscriber subscriber : subscribers) {
+              notifications.add(new NotificationContent.Builder(subscriber.firstName, subscriber.lastName, subscriber.email, 
+              title, projectCode, failedQCCount, availableDataCount).build())
+            }
+          }
+          output.createdNotifications(notifications)
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:59]: This looks very complex. You could improve code readability if you extract this into another function. 
Effectively you collect the information for each project here, right?
You could introduce a method that generates the notification for a given project, and call this method for each project.

Also the association with the subscriber should be done outside of this method. If you would like to discuss some more detaild suggestions let me know :)

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/FetchSubscriberDbConnector.groovy [hunk 7]
```diff
@@ -60,7 +61,35 @@ class FetchSubscriberDbConnector implements FetchSubscriberDataSource{
...
+            connection.withCloseable { Connection con ->
+                    String sqlQuery = "select openbis_project_identifier, short_title from projects"
+
+                    PreparedStatement preparedStatement = con.prepareStatement(sqlQuery)
+                    preparedStatement.execute()
+                    def resultSet = preparedStatement.getResultSet()
 
+                    while(resultSet.next()){
+                        String identifier = resultSet.getString("openbis_project_identifier")
+                        String code = identifier.substring(identifier.length() - 5)
```
Comment (jenniferboedker) [inline sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/FetchSubscriberDbConnector.groovy:81]: why -5?

Comment (jenniferboedker) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/FetchSubscriberDbConnector.groovy:81]: bzw how does an identifier look like?

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 8]
```diff
@@ -15,156 +19,100 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
-                listProjectsBuffer << "\n"
-                listProjectsBuffer.append(sampleCode)
-                listProjectsBuffer << "\n"
+    
+    private Map<String, List<String>> getProjectsWithSamples(List<String> samples) {
+        Map<String, List<String>> projectToSamples = new HashMap<>()
+        samples.each { sample ->
+            String project = sample.substring(0, 5)
+            if(projectToSamples.containsKey(project)) {
+              def list = [sample]
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:83]: can you think of a better variable name?

Comment (jenniferboedker) [reply sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:83]:
```suggestion
              def samples = [sample]
```

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 9]
```diff
@@ -15,156 +19,100 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
-        fetchSubscriberInput.fetchSubscriber(date)
-        //assume notificationPerSubscriber will be filled by the output method fetchedSubscribers of the FetchSubscriber use case
-        output.createdNotifications(notificationPerSubscriber)
+      LocalDate localDate = LocalDate.parse(date)
+      
+      try {
+          //1. get todays notifications
+          Map<String, Status> sampleToStatus = dataSource.getUpdatedSamplesForDay(localDate)
+          //2. get projects
+          Map<String,List<String>> projectsWithSamples = getProjectsWithSamples(sampleToStatus.keySet().asList())
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:41]: Do we need all samples per project?
Can we skip this part and directly count and associate the number of samples for a project?
Just one method called getSamplesPerStatus(String project, Status status) or so?

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/NotificationContent.groovy [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * <h1>A DTO containing the fields required in the email</h1>
+ *
+ * <p>The content of this class is based on all the fields that need be part of the final email send to the customer</p>
+ *
+ * @since 1.0.0*
+ */
+
+@EqualsAndHashCode
+class NotificationContent {
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/NotificationContent.groovy:14]: Very nice idea, i like that!

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/NotificationContent.groovy [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+    /*Customer Information */
+    final String customerFirstName
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/NotificationContent.groovy:17]: maybe add the customer title here too

File: sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy [hunk 12]
```diff
@@ -14,9 +14,68 @@ import java.time.LocalDate
...
+//        and: "Datasource that returns the Subscriber list and the updated Sample Status"
+//        ds.getUpdatedSamplesForDay(_ as LocalDate) >> updatedSamples
+//        ds.getSubscriberForProject(_ as String) >> subscribers
+//
+//        when: "The CreateNotification use case is triggered"
+//        createNotification.createNotifications("2020-08-17")
+//
+//        then: "A map associating the notifications with the subscriber for the provided date is returned"
+//        1* output.createdNotifications(_ as Map<Subscriber, String>)
+//    }
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy:74]: Still needed?

```suggestion
```

**Review by jenniferboedker — CHANGES_REQUESTED** [review]
Thanks a lot for contributing to this project! I like your suggestion, especially the introduction of the DTO.

I gave you some feedback concerning code readability and I was wondering why you introduced the FetchUpdateDataSource. Maybe you can comment on that?

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy (hunk 4 above)
Comment (jenniferboedker) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:26]: maybe @Steffengreiner can comment on that?:)

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/FetchSubscriberDbConnector.groovy (hunk 7 above)
Comment (wow-such-code) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/FetchSubscriberDbConnector.groovy:81]: identifier = /space/projectCode (which is 5 symbols long)
It's the last 5 symbols, didn't want to perform splits, which might lead to array out of bounds exceptions

File: sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy (hunk 12 above)
Comment (wow-such-code) [reply sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy:74]: a placeholder for now, want to add more tests once I can run them

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy (hunk 4 above)
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:26]: I think this will be where the dependency manager will get the List of notificationContent needed for the email generation

Comment (jenniferboedker) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:26]: ok, but is this going to be implemented with your PR?

File: sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy (hunk 12 above)
Comment (jenniferboedker) [reply sample-notificator-domain/src/test/groovy/life/qbic/business/subscription/CreateNotificationSpec.groovy:74]: I would prefer to delete it since it is just a copy of the test below, right?

File: sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy (hunk 4 above)
Comment (Steffengreiner) [reply sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/components/CreateNotificationPresenter.groovy:26]: Yes I'll adapt the fill PR #17 as soon as this is merged to use this list 👍

**Comment by wow-such-code** [comment]
notification tests now work, there are some problems with the other use case tests, but I'm not sure this is needed anymore, anyway

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 13]
```diff
@@ -22,56 +20,58 @@ import java.time.LocalDate
...
-            
-            //4. get subscribers of this projects
-            List<Subscriber> subscribers = dataSource.getSubscriberForProject(projectCode)
-            
-            for(Subscriber subscriber : subscribers) {
-              notifications.add(new NotificationContent.Builder(subscriber.firstName, subscriber.lastName, subscriber.email, 
-              title, projectCode, failedQCCount, availableDataCount).build())
-            }
-          }
+          updatedSamplesWithStatus = dataSource.getUpdatedSamplesForDay(localDate)
```
Comment (KochTobi) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy]: Where do we filter for failed QC?

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 14]
```diff
@@ -22,56 +20,58 @@ import java.time.LocalDate
...
       }
     }
+
+    private List<NotificationContent> createNotificationForProject(String projectCode, List samples){
+        List<NotificationContent> notifications = []
+        //3. get project names
+        Map<String,String> projectsWithTitles = dataSource.fetchProjectsWithTitles()
+
+        int failedQCCount = filterSamplesByStatus(samples, updatedSamplesWithStatus, "SAMPLE_QC_FAIL").size()
+        int availableDataCount = filterSamplesByStatus(samples, updatedSamplesWithStatus, "DATA_AVAILABLE").size()
```
Comment (jenniferboedker) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy]: Here we filter for the status

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 15]
```diff
@@ -15,156 +16,89 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
+
+
+    private Set<Project> getProjects() {
+        Set<Project> projects = new HashSet<>()
+
+        updatedSamplesWithStatus.keySet().each { sample ->
+            Project project = new Project()
+            project.code = sample.substring(0, 5)
+
+            Optional<Project> found = projects.stream().filter({it.code == project.code}).findFirst()
```
Comment (KochTobi) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:93]:
```suggestion
            Optional<Project> foundProject = projects.stream().filter({it.code == project.code}).findFirst()
```
NIT: This way it is clear what this Optional contains without looking into the declaration :)

Comment (KochTobi) [reply sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:93]:
```suggestion
            Optional<Project> found = projects.stream().filter({it == project}).findFirst()
```
As the Project implements equality with the code only. This is another possibility. However, I think your solution is better here since it does not depend on the implementation logic of the equals method :+1:

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy [hunk 16]
```diff
@@ -15,156 +16,89 @@ import life.qbic.business.subscription.fetch.FetchSubscriberOutput
...
+        List<String> filteredCodes = samples.stream().filter(isOfStatus).collect().toList()
+
+        return filteredCodes
+    }
+
+
+    private Set<Project> getProjects() {
+        Set<Project> projects = new HashSet<>()
+
+        updatedSamplesWithStatus.keySet().each { sample ->
```
Comment (KochTobi) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/CreateNotification.groovy:89]:
```suggestion
        updatedSamplesWithStatus.keySet().each { sampleCode ->
```
Maybe it would be an idea to rename the variable to make it clear that we only have the sample code and not another representation of the sample here.

File: sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/FetchProjectDataSource.groovy [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+import life.qbic.business.subscription.Subscriber
+
+import java.time.LocalDate
+
+/**
+ * <p>Fetches the project from the database</p>
+ *
+ * @since 1.0.0
+ */
+interface FetchProjectDataSource {
```
Comment (KochTobi) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/notification/create/FetchProjectDataSource.groovy:14]: :+1:

File: sample-notificator-domain/src/main/groovy/life/qbic/business/subscription/fetch/FetchSubscriber.groovy [hunk 18]
```diff
@@ -53,7 +53,7 @@ class FetchSubscriber implements FetchSubscriberInput {
 
     /**
      * Collects all updated samples (from all projects) for the projects a user has subscribed to.
-     * @return A map with the user id (email address) and the updated samples of his subscribed projects
+     * @return A map with the user id (email address) and the updated samples of their subscribed projects
```
Comment (KochTobi) [inline sample-notificator-domain/src/main/groovy/life/qbic/business/subscription/fetch/FetchSubscriber.groovy:56]:
```suggestion
     * @return a list of subscribers with subscriptions included
```
I think this is no longer a map 😉

**Review by KochTobi — APPROVED** [review]
Some nit-picking. Otherwise, looks good to me :+1:
